### PR TITLE
Use the 'no outstanding' rule to defer idle timeout

### PR DESCRIPTION
**Pull request by martinthomson**
Closes #1648.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2499,9 +2499,11 @@ A connection that remains idle for longer than the advertised idle timeout (see
 draining state when the idle timeout expires.
 
 Each endpoint advertises their own idle timeout to their peer. The idle timeout
-starts from the last packet received, or the first packet sent after the last
-received packet.  The latter condition ensures that initiating new activity
-postpones a timeout.
+starts from the last packet received.  In order to ensure that initiating new
+activity postpones an idle timeout, an endpoint restarts this timer when sending
+a packet if there are no other packets outstanding.  This condition only applies
+to packets that contain other than ACK or PADDING frames; those packets could
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: "that contain other" is awkward, how about "that contain frames other"

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2499,9 +2499,11 @@ A connection that remains idle for longer than the advertised idle timeout (see
...
+remain outstanding longer and prevent the timeout from being extended.
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: How about a separate sentence "Packets that contain only ACK or PADDING frames may never be acknowledged, so they do not extend the timeout."

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -2499,9 +2499,11 @@ A connection that remains idle for longer than the advertised idle timeout (see
 draining state when the idle timeout expires.
 
 Each endpoint advertises their own idle timeout to their peer. The idle timeout
-starts from the last packet received, or the first packet sent after the last
-received packet.  The latter condition ensures that initiating new activity
-postpones a timeout.
+starts from the last packet received.  In order to ensure that initiating new
+activity postpones an idle timeout, an endpoint restarts this timer when sending
+a packet if there are no other packets outstanding.  This condition only applies
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: How about "In order to ensure that initiating new activity postpones an idle timeout, an endpoint restarts this timer when sending a packet containing frames besides ACK or PADDING and no other packets are outstanding."

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: This comment is obsolete if you make the above change or something similar.

**Comment by ianswett** [comment]
On a high level, there's some remaining vagueness about what outstanding means.  Particularly now that PADDING frames count towards bytes in flight, but do not elicit an ACK.

**Comment by mikkelfj** [comment]
Surely an ACK only packet does not prevent idle timeout from refreshing? It merely does not controbute to progress?

**Comment by martinthomson** [comment]
Receiving an ACK-only packet will reset the timer.  Sending one might.  If there are no other outstanding packets, then an ACK is sent after receiving a packet and after waiting for any ACK delay.  That will refresh the timer.  However, you won't send another ACK-only packet until a packet is declared lost, which only happens when new packets arrive.

**Comment by mikkelfj** [comment]
I'm just find the word "prevent" confusing here:

"Packets that contain only ACK or
 PADDING frames are not acknowledged until an endpoint has other frames to send,
 so they could prevent the timeout from being refreshed."

**Comment by martinthomson** [comment]
An ACK- and PADDING-only packet would not get acknowledged and so remain unacknowledged indefinitely.  If the rule is "don't refresh if you have unacknowledged packets", then those packets would prevent the timeout from being restarted when a new packet is sent.

**Comment by mikkelfj** [comment]
I see - but why don't you exclude those packets. Simplicity?

**Pull request merged**